flash/image/classification/integrations/baal.py

```diff
@@ -0,0 +1,143 @@
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+from copy import deepcopy
+from functools import partial
+from typing import Any, Callable, Dict, Optional
+
+import numpy as np
+import torch
+from pytorch_lightning.loops import Loop
+from pytorch_lightning.loops.fit_loop import FitLoop
+from pytorch_lightning.trainer.progress import Progress
+from pytorch_lightning.trainer.states import RunningStage, TrainerFn, TrainerStatus
+from torch.nn.modules.batchnorm import _BatchNorm
+
+import flash
+from flash.core.utilities.imports import _BAAL_AVAILABLE, requires
+
+if _BAAL_AVAILABLE:
+    from baal.active import ActiveLearningDataset
+    from baal.active.heuristics import AbstractHeuristic, BALD
+
+
+class ActiveLearningTrainer(flash.Trainer):
+    def __init__(self, *args, **kwags):
+        super().__init__(*args, **kwags)
+
+        active_learning_loop = ActiveLearningLoop(label_epoch_frequency=1, heuristics=BALD(reduction=np.mean))
+        active_learning_loop.connect(self.fit_loop)
+        self.fit_loop = active_learning_loop
+        active_learning_loop.trainer = self
+
+
+class ActiveLearningLoop(Loop):
+    @requires("baal")
+    def __init__(self, label_epoch_frequency: int, heuristics: "AbstractHeuristic", inference_iteration: int = 2):
+        super().__init__()
+        self.label_epoch_frequency = label_epoch_frequency
+        self.fit_loop: Optional[FitLoop] = None
+        self.state_dict: Optional[Dict[str, torch.Tensor]] = None
+        self.progress = Progress()
+        self.heuristics = heuristics
+        self._should_continue: bool = False
+        self.inference_iteration = inference_iteration
+
+    def __getattr__(self, key):
+        if key not in self.__dict__:
+            return getattr(self.fit_loop, key)
+        return self.__dict__[key]
+
+    def connect(self, fit_loop: FitLoop):
+        self.fit_loop = fit_loop
+        self.max_epochs = self.fit_loop.max_epochs
+        self.fit_loop.max_epochs = self.label_epoch_frequency
+
+    def _to_active_dataset(self, dataset):
+        return ActiveLearningDataset(dataset, labelled=torch.arange(len(dataset) - 10))
+
+    def on_run_start(self, *args: Any, **kwargs: Any) -> None:
+        self.trainer.predict_loop._return_predictions = True
+        self.state_dict = deepcopy(self.trainer.lightning_module.state_dict())
+        if self.trainer.datamodule is not None:
+            self.trainer.datamodule.register_dataset_hooks(self._to_active_dataset, RunningStage.TRAINING)
+            self.trainer.datamodule.register_dataset_hooks(self._to_active_dataset, RunningStage.VALIDATING)
+            self.trainer.datamodule.register_dataset_hooks(self._to_active_dataset, RunningStage.TESTING)
+            self.trainer.datamodule.register_dataset_hooks(self._to_active_dataset, RunningStage.PREDICTING)
```

Comment:
normally, we do not wrap the data used for validation and test with `ActiveLearningDataset`, as these sets are supposed to be fully labelled. Is there a reason that we wrap all of them with `ActiveLearningDataset` here? 
On the other hand, Do we expect the prediction set to be the `pool` or separate?

Reply:
To add to @parmidaatg's comment, it would be great to have the option to turn-on "dynamic" validation set: the validation set contains the same number of instances as the current training budget. This mimics a more realistic active learning scenario

Reply:
Thanks @parmidaatg @pietrolesci,

@parmidaatg Great point, there is no reason for them to get wrapped. I was just debugging the code :)
Only predict_dataset would be wrapped into an ActiveLearningDataset as it contains the non-labelled data.

@pietrolesci I would be curious if you could share more details on how dynamic validation set is usually implemented.

Reply:
@tchaton one way I did it in the past was to add to the validation set the same number of instances added to the training set, chosen at random.

I must add, in the future one can also think to devise better strategies to actively select even the validation set (think of the applications in meta-learning), but this is another discussion :)

Reply:
@tchaton to address your reply to @parmidaatg - do you think there is a clean way to allow the user to choose what to wrap?

Reply:
true, we haven't supported the automated selection of validation-set in BaaL yet. I believe it should be an open issue which is not a priority at the time but potentially it would be nice to support having the same number of examples in validation set or maybe the same potion as the labelled data vs pool.